spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -213,21 +214,50 @@ public Object postProcessBeforeInitialization(Object bean, String beanName) thro
 	}
 
 	@Override
-	public Object postProcessAfterInitialization(final Object bean, String beanName) throws BeansException {
+	public Object postProcessAfterInitialization(final Object bean, final String beanName) throws BeansException {
 		Class<?> targetClass = AopUtils.getTargetClass(bean);
+		final RabbitListener classLevelListener = AnnotationUtils.findAnnotation(bean.getClass(), RabbitListener.class);
+		final List<Method> multiMethods = new ArrayList<Method>();
 		ReflectionUtils.doWithMethods(targetClass, new ReflectionUtils.MethodCallback() {
+
 			@Override
 			public void doWith(Method method) throws IllegalArgumentException, IllegalAccessException {
 				RabbitListener rabbitListener = AnnotationUtils.getAnnotation(method, RabbitListener.class);
 				if (rabbitListener != null) {
-					processAmqpListener(rabbitListener, method, bean);
+					processAmqpListener(rabbitListener, method, bean, beanName);
+				}
+				if (classLevelListener != null) {
+					RabbitHandler rabbitHandler = AnnotationUtils.getAnnotation(method, RabbitHandler.class);
+					if (rabbitHandler != null) {
```

Comment:
This is not Spring Integration - I think it's important to be able to specify which methods are candidates - to be honest, I really don't like it that SI considers all public methods candidates - it can causes problems when unintended methods are considered.

Further, as I discussed with @snicoll, in future, we might add more attributes to `@RabbitHandler` to assist with method matching (e.g. add a SpelExpression `matchFor = "headers['foo'] == 'bar'"` which could allow enhancement to matching instead of just matching on payload type).